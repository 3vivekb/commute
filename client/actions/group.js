import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addGroup = createAction('add group')
const appendCommuters = createAction('append commuters')
export const appendToGroup = ({ newCommuters, groupId }) => {
  newCommuters = newCommuters.map((commuter) => {
    commuter.groupId = groupId
    return commuter
  })
  return [
    appendCommuters({ commuters: newCommuters, groupId }),
    push(`/group/${groupId}`)
  ] // TODO save to server
}
export const createGroup = (newGroup) => {
  newGroup.id = uuid.v4()
  newGroup.allAddressesGeocoded = false
  if (!newGroup.commuters) {
    newGroup.commuters = []
  } else {
    newGroup.commuters = newGroup.commuters.map((commuter) => {
      commuter.groupId = newGroup.id
      return commuter
    })
  }
  return [
    addGroup(newGroup),
    push(`/group/${newGroup.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete group')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/group/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteGroup = ({ id, organizationId }) => [
  deleteLocally({ id, organizationId }),
  push(`/organization/${organizationId}`)
]

const updateLocally = createAction('set group')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/group/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateGroup = (group) => [
  updateLocally(group)
]
