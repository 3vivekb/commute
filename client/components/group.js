import React, {Component} from 'react'

export default class Group extends Component {
  static propTypes = {}

  render () {
    return <h1>Group</h1>
  }
}
