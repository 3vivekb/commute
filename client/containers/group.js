import {connect} from 'react-redux'

import Group from '../components/group'

function mapStateToProps (state) {
  return {}
}

export default connect(mapStateToProps)(Group)
