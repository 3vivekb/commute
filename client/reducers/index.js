import {handleActions} from 'redux-actions'

import * as commuter from './commuter'
import * as multiSite from './multi-site'
import * as site from './site'
import * as user from './user'

export default {
  commuter: handleActions(commuter.reducers, commuter.initialState),
  multiSite: handleActions(multiSite.reducers, multiSite.initialState),
  site: handleActions(site.reducers, site.initialState),
  user: handleActions(user.reducers, user.initialState)
}
