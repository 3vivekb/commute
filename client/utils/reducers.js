import update from 'react-addons-update'

import {addToEntityMap, addEntitiesToEntityMap, deleteFromEntityMap, deleteManyFromEntityMap} from './entities'

function addChildEntityToParent (parentMap, newEntity, parentIdField, childPluralName) {
  return update(parentMap, {
    [newEntity[parentIdField]]: {
      [childPluralName]: {
        $push: [newEntity._id]
      }
    }
  })
}

/**
 * Make handlers for children creation and deletion events
 *
 * @param  {Object} cfg  An configuration object structured as follows:
 *   - childPluralName (String): Plural name of child (parent's child id array)
 *   - childSingularName (String): Singular name of child (for making reducer title)
 *   - parentIdField (String): Id field of parent in child entity
 * @return {Object}     Reducers
 */
export function makeChildrenHandlers (cfg) {
  const reducers = {}
  reducers[`add ${cfg.childSingularName}`] = function (state, action) {
    const newEntity = action.payload
    return addChildEntityToParent(
      state,
      newEntity,
      cfg.parentIdField,
      cfg.childPluralName
    )
  }

  reducers[`add many ${cfg.childPluralName}`] = function (state, action) {
    const newEntities = action.payload
    let updatedState = state
    newEntities.forEach(newEntity => {
      updatedState = addChildEntityToParent(
        updatedState,
        newEntity,
        cfg.parentIdField,
        cfg.childPluralName
      )
    })
    return updatedState
  }

  reducers[`delete ${cfg.childSingularName}`] = function (state, action) {
    const entity = action.payload
    const childIds = state[entity[cfg.parentIdField]][cfg.childPluralName]
    return update(state, {
      [entity[cfg.parentIdField]]: {
        [cfg.childPluralName]: {
          $set: childIds.filter((id) => id !== entity._id)
        }
      }
    })
  }

  return reducers
}

/**
 * Make generic reducers for an entity type
 *
 * @param  {Object} cfg  An object with name and handlers parameters structured as follows:
 *   - name: must have singular and plural keys with the corresponding strings
 *   - handlers: An array of strings of handlers.  Possible values: ['add', 'add many', 'delete', 'set', 'set many']
 * @return {Object}      The generated reducers
 */
export function makeGenericReducerHandlers (cfg) {
  const reducers = {
    'log out' (state, action) {
      return {}
    }
  }
  if (cfg.handlers.indexOf('add') !== -1) {
    reducers[`add ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('add many') !== -1) {
    reducers[`add many ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('delete') !== -1) {
    reducers[`delete ${cfg.name.singular}`] = function (state, action) {
      return deleteFromEntityMap(state, action.payload._id)
    }
  }

  if (cfg.handlers.indexOf('delete many') !== -1) {
    reducers[`delete many ${cfg.name.plural}`] = function (state, action) {
      return deleteManyFromEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set') !== -1) {
    reducers[`set ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set many') !== -1) {
    reducers[`set ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  return reducers
}
