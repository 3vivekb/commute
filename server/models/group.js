const {Schema} = require('mongoose')

const Commuter = require('./commuter')

const groupSchema = new Schema({
  name: {
    required: true,
    type: String
  },
  organizationId: {
    ref: 'Organization',
    required: true,
    type: Schema.Types.ObjectId
  }
})

groupSchema.pre('remove', function (next) {
  // CASCADE DELETE
  Commuter.remove({ groupId: this._id }).exec()
  next()
})

module.exports = groupSchema
