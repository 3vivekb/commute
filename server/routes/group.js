const pick = require('lodash.pick')

const models = require('../models')
const makeRestEndpoints = require('./').makeRestEndpoints

module.exports = function makeRoutes (app) {
  makeRestEndpoints(app,
    {
      childModels: [{
        foreignKey: 'groupId',
        key: 'commuters',
        model: models.Commuter
      }],
      commands: {
        'Collection GET': {},
        'GET': {},
        'DELETE': {},
        'PUT': {}
      },
      model: models.Group,
      name: 'group'
    }
  )

  // rest-ish endpoints
  app.post('/api/group', (req, res) => {
    res.set('Content-Type', 'application/json')
    const handleErr = (err) => res.status(500).json({error: err})
    let body = req.body
    if (!Array.isArray(body)) {
      body = [body]
    }
    const groupsToCreate = body.map((newGroup) => pick(newGroup, ['name', 'organizationId']))
    models.Group.create(groupsToCreate)
      .then((groups) => {
        const createGroupCommutersPromises = groups.map((group, idx) => {
          return new Promise((resolve, reject) => {
            if (body[idx].commuters && body[idx].commuters.length > 0) {
              // insert commuters one-by-one to trigger pre-save hook
              Promise.all(body[idx].commuters.map((commuter) => {
                return models.Commuter.create(Object.assign(commuter, { groupId: group._id }))
              }))
                .then((data) => {
                  const output = Object.assign({ commuters: data.map((commuter) => commuter._id) }, group._doc)
                  resolve(output)
                })
                .catch(reject)
            } else {
              resolve(group)
            }
          })
        })
        Promise.all(createGroupCommutersPromises)
          .then((data) => {
            res.json(data)
          })
          .catch((err) => {
            handleErr(err)
          })
      })
      .catch(handleErr)
  })
}
