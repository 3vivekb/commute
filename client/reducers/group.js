import update from 'react-addons-update'

import {addToEntityMap, entityArrayToEntityIdArray} from '../utils/entities'
import {makeChildrenHandlers, makeGenericReducerHandlers} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['delete', 'set', 'set many'],
  name: {
    singular: 'group',
    plural: 'groups'
  }
})

reducers['add group'] = function (state, action) {
  const newGroup = {...action.payload}
  newGroup.commuters = newGroup.commuters.map((commuter) => commuter._id)
  return addToEntityMap(state, newGroup)
}

reducers['append commuters'] = function (state, action) {
  const affectedGroup = state[action.payload.groupId]
  const modifiedGroup = update(affectedGroup, {
    commuters: {
      $push: entityArrayToEntityIdArray(action.payload.commuters)
    }
  })
  return addToEntityMap(state, modifiedGroup)
}

Object.assign(reducers, makeChildrenHandlers({
  childPluralName: 'commuters',
  childSingularName: 'commuter',
  parentIdField: 'groupId'
}))

export const initialState = {}
