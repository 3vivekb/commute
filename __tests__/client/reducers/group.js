/* globals describe, expect, it */

import {handleActions} from 'redux-actions'

import {commuterSal, mockGroup, mockGroupCreation, mockStores} from '../../test-utils/mock-data'
import {makeChildrenHandlerTestCases, makeGenericReducerTestCases} from '../../test-utils/reducers'

import * as group from '../../../client/reducers/group'

describe('client > reducers > group', () => {
  const anotherGroup = { _id: 'group-3', name: 'Another Group', commuters: [] }

  makeChildrenHandlerTestCases({
    add: {
      affectedParentId: 'group-2',
      initialState: mockStores.withAnalysisRun.group,
      payload: { _id: 'commuter-new', groupId: 'group-2' }
    },
    childPluralName: 'commuters',
    childSingularName: 'commuter',
    delete: {
      affectedParentId: 'group-2',
      initialState: mockStores.withAnalysisRun.group,
      payload: { _id: 'commuter-2', groupId: 'group-2' }
    },
    initialState: group.initialState,
    reducers: group.reducers
  })

  makeGenericReducerTestCases({
    handlers: {
      delete: {
        initialState: mockStores.withAnalysisRun.group,
        payload: mockGroup
      },
      set: {
        initialState: mockStores.withBlankOrganization.group,
        payload: anotherGroup
      },
      'set many': {
        initialState: group.initialState,
        payload: [anotherGroup]
      }
    },
    initialState: group.initialState,
    name: {
      plural: 'groups',
      singular: 'group'
    },
    reducers: group.reducers
  })

  it('should handle add group', () => {
    const reducer = handleActions(group.reducers, group.initialState)
    const newGroup = {...mockGroupCreation}
    const action = { payload: newGroup, type: 'add group' }
    const result = reducer(mockStores.withBlankOrganization.group, action)
    expect(result['new-group'].name).toEqual('Fake Group')
    expect(result).toMatchSnapshot()
  })

  it('should handle append commuters', () => {
    const reducer = handleActions(group.reducers, group.initialState)
    const action = {
      type: 'append commuters',
      payload: {
        groupId: 'group-2',
        commuters: [commuterSal]
      }
    }
    const result = reducer(mockStores.withAnalysisRun.group, action)
    expect(result['group-2'].commuters.length).toBe(2)
    expect(result).toMatchSnapshot()
  })
})
