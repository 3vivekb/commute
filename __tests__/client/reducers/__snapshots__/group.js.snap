exports[`client > reducers > group commuters children handler reducer tests it should handle add commuter 1`] = `
Object {
  "group-2": Object {
    "allAddressesGeocoded": true,
    "commuters": Array [
      "commuter-2",
      "commuter-new",
    ],
    "id": "group-2",
    "name": "Mock Group",
    "organizationId": "organization-2",
  },
}
`;

exports[`client > reducers > group commuters children handler reducer tests it should handle delete commuter 1`] = `
Object {
  "group-2": Object {
    "allAddressesGeocoded": true,
    "commuters": Array [],
    "id": "group-2",
    "name": "Mock Group",
    "organizationId": "organization-2",
  },
}
`;

exports[`client > reducers > group generic reducer tests should handle delete group 1`] = `Object {}`;

exports[`client > reducers > group generic reducer tests should handle set group 1`] = `
Object {
  "group-3": Object {
    "commuters": Array [],
    "id": "group-3",
    "name": "Another Group",
  },
}
`;

exports[`client > reducers > group generic reducer tests should handle set groups 1`] = `
Object {
  "group-3": Object {
    "commuters": Array [],
    "id": "group-3",
    "name": "Another Group",
  },
}
`;

exports[`client > reducers > group should handle add group 1`] = `
Object {
  "new-group": Object {
    "commuters": Array [
      "new-commuter",
    ],
    "id": "new-group",
    "name": "Fake Group",
    "organizationId": "organization-1",
  },
}
`;

exports[`client > reducers > group should handle append commuters 1`] = `
Object {
  "group-2": Object {
    "allAddressesGeocoded": true,
    "commuters": Array [
      "commuter-2",
      "commuter-3",
    ],
    "id": "group-2",
    "name": "Mock Group",
    "organizationId": "organization-2",
  },
}
`;
